### DISCUSSION TITLE
[Guide] Separating code into modular files and using 'include' to load them

### DISCUSSION THREAD

**Discussion by cbot-coder** [colobot/colobot#1373] [Workshop]
## Background and Basic Coding Principles

A good programming practice is to break up larger programs into smaller modules that can be managed more easily. They can also package code that is used frequently and in different programs, so that you don't need to copy the same code over and over again into your new programs. Over time, you can build up a "library" of code that you build upon so you don't need to reinvent the wheel every time. Most modern languages provide some support for separating code into modules or files, and then loading them into your main program using an `include` statement or something similar.

## Problem: CBOT does not support `include`

The CBOT language has no native concept of `include` for code files. This is one of the biggest pains in doing more advanced CBOT coding: you typically have to keep all the code in one file. Yes, Colobot's built-in IDE allows you to load different pieces of code manually into different program slots on the left, and they can call each other, but there are at least two problems with this:

1. When stepping through code to debug, the code window can only show you the code in the current program slot. It doesn't show you what is happening in the code it calls in any other slots. (This is also a problem when two or more bots share the same code, but I haven't found a good solution for that yet.)
2. It's super inconvenient, especially as your library of code grows. Every time you change a library file, you have to remember to load it separately again into its slot.

## Solution: Use the C Preprocessor

C/C++ handles `include` using what is called a "preprocessor" program. It "pre-processes" code files before compilation, loading code from `include`d files into the main program and even supporting simple "macros" (shortcuts or function-like code replacements), among other things. I realized that CBOT is somewhat like Java and somewhat like C++, and so the C preprocessor works almost perfectly for CBOT code. There is a bit of tweaking still needed, so I wrote myself a short wrapper script to invoke the C preprocessor on `.cbot` files (the file extension for my CBOT code files).

**CAVEAT**: I'm a Linux nerd, and this only works on Linux as far as I know. If anyone is using Mac or Windows and would like to post their version or solution, please do!

Here is the script, named `cbot-pp`:

```bash
#!/bin/sh

# Preprocessor for CBOT code (Colobot code, a simple language that
# resembles C++ and Java).
#
# This uses 'cpp' but passes some default options we like, and removes
# the GPL header in the output

# options: preserve comments, don't output linemarkers, keep whitespace
OPTIONS="-C -P -traditional"
CPP_CMD="cpp"
# I don't know how else to strip out the comments that Gnu cpp inserts
# NOTE! Other versions may require tweaking this filter
SED_FILTER='/Copyright (C) .* Free Software Foundation/,/Zanabazar/ d; /./,$ p'

IN_FILE="$1"; shift
OUT_FILE="$1"

if [ "$OUT_FILE" != "" ]; then
  $CPP_CMD $OPTIONS "$IN_FILE" | sed -n "$SED_FILTER" > "$OUT_FILE"
else
  $CPP_CMD $OPTIONS "$IN_FILE" | sed -n "$SED_FILTER"
fi
```

## Example

Now, I can write libraries of code and structure them elegantly into different modules and classes, and only include the ones I need in my per-mission solution files!

Here's a simple example from Earth 2, the earliest mission that can be programmed. It shows a very early version of some of my code. I have a solution file, and a separate file containing a `makeBuilding` function.

#### `1Earth2Solution.cbot`
```c++
extern void object::Earth2Solution() {
  makeBuilding(ResearchCenter, "Research Center");
  makeBuilding(BotFactory, "Bot Factory");
}

#include "MakeBuilding.cbot"
```

#### `MakeBuilding.cbot`
```c++
// Makes a building, if it's not built already. Parameters:
// - category: the building category
// - name: the name of the building, for user messages
// - botBuild: whether the bot should build it, or prompt the
//             astronaut to build it instead
// Returns:
// - true if this call made the building
// - false if the building already existed
public boolean makeBuilding(int category, string name,
                            boolean botBuild = true) {
  object item = radar(category);
  if (item == null) {
    message("We need a " + name);
	message("Finding a Titanium cube...");
    item = radar(Titanium);
    goto(item.position);

    // The bot can build it, if requested...
    if (botBuild) build(category);
    
    item = radar(category);
    if (item != null) {
      message("Built a " + name);
    } else {
      // Ask the astronaut to build it
      move(-10);
      while((item = radar(category)) == null) {
        message("There is a titanium cube in front of me.  Please build a " + name);
        wait(10);
      }
    }      
    return true;
  }
  return false; // nothing built, already there
}
```

These are two separate files, but to combine them, I run:

```bash
cbot-pp 1Earth2Solution.cbot ~/.local/share/colobot/savegame/Player/program/1Earth2Sol.txt
```

Then, I start the mission, select the WheeledGrabber bot, create a new program slot...
![image](https://user-images.githubusercontent.com/75964590/104200443-49c68180-53ee-11eb-83f1-14e2520fd8f3.png)

...open the program I outputted from the `cbot-pp` command...
![image](https://user-images.githubusercontent.com/75964590/104200694-93af6780-53ee-11eb-8e36-c7d55b024a98.png)

...and there it is, both files combined into one program!
![image](https://user-images.githubusercontent.com/75964590/104200845-ba6d9e00-53ee-11eb-80ee-9f9bb8814ad7.png)

Of course, running it is the proof of success:
![image](https://user-images.githubusercontent.com/75964590/104200992-e2f59800-53ee-11eb-9c09-a2021abb4598.png)

And without any further manual action on your part, the bot completes the mission!
![image](https://user-images.githubusercontent.com/75964590/104201091-fc96df80-53ee-11eb-9337-5d051b7c5148.png)

## Problem 2: What if a file gets `include`d twice?

This frequently occurs once your library of code gets larger, and files start including other files, which include other files. You may have something like this:

##### `Mission.cbot`
```c++
// code to solve the mission...
// it references some constants in Constants.cbot:
#include "Constants.cbot"
// and it also uses a class to handle common things with bots:
#include "Bot.cbot"
```

##### `Bot.cbot`
```c++
// code to handle common things with bots
// it also needs to reference some constants in Constants.cbot:
#include "Constants.cbot"
```

##### `Constants.cbot`
```c++
// some commonly-used constants are defined here...
```

The problem with this is that `Constants.cbot` will get `include`d twice: once by the main program, and again by the `Bot.cbot` file. What is the solution? Again, we use the solution from C/C++: preprocessor directives. These are special instructions that begin with `#` and are directed to the preprocessor only, telling it how to handle code. One of the most common is the `#define` directive. The most simple use is to define a variable, basically to act as a flag. Let's use this at the beginning of our `Constants.cbot` file:

```c++
#define CONSTANTS_LOADED
```

Then, we wrap the entire file within the `#ifndef` ... `#endif` directive pair ("if not defined" ... "end") to avoid including the entire file code if that flag has already been defined (if the file has already been loaded). The final file looks like this:

##### `Constants.cbot`
```c++
#ifndef CONSTANTS_LOADED
#define CONSTANTS_LOADED
// some commonly-used constants are defined here...
#endif
```

The first time the file is `include`d, it will define CONSTANTS_LOADED and then include the contents. The second time, it sees that CONSTANTS_LOADED is defined already, so it skips including everything inside that file. Solved!

## Advanced: Preprocessor Macros

I wouldn't encourage using these too much, because they can get hairy and complicated. However, for simple things, you can use `#define` to define certain constants or even simple functions that you want the preprocessor to inline by substitution into the code whenever you reference them. A few simple examples:

```c++
#define PI 3.14159
#define DOUBLE(x) ((x) * 2.0)
#define MAX(x, y) (((x) > (y)) ? (x) : (y))
// The extra parentheses in the #define above are necessary so that
// when we pass in complex expressions, they don't mess up unexpectedly
// due to operator precedence rules.

// Now in my CBOT code, whenever I write the following:
float circumference = DOUBLE(radius) * PI;
// the preprocessor actually outputs the following:
float circumference = ((radius) * 2.0) * 3.14159;

// Similarly, if I write:
int a = 1, b = 2;
int highest = MAX(a + 2, b);
// the preprocessor outputs:
int a = 1, b = 2;
int highest = (((a + 2) > (b)) ? (a + 2) : (b));
```

**Comment by Lars0312** [comment]
Hello @cbot-coder,

I have test ur code on my mac Machine and its work very well.
Im not so fimiliar with c++ or cli compiler.
I have make a run.task / build.task for Visual Studio Code to run ur Implementation.

`{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            "label": "Build cbot",
            "type": "shell",
            // ../${fileBaseNameNoExtension} is only the Active name from the Opened File
            // i should ever be the main solution
            "command": "sh cbot-pp.sh ${fileBasename} ../${fileBasenameNoExtension}.txt",
            "problemMatcher": [],
            "group": {
                "kind": "build",
                "isDefault": true
            }
        }
    ]
}`

maybe it help other people.

I would love to here your and others feedback.

Greetings and good night.

**Comment by Three-Search** [comment]
I copied my SatCom.cbot library file into colobot\data\ai folder, then I modified scene.txt of each level mission to add LevelController object as below.
BeginObject
LevelController script="SatCom.cbot"

My bot can call any function in my SatCom.cbot.
